### PR TITLE
Directly: Add state to Redux

### PR DESCRIPTION
**Pull request by mattwondra**
This PR introduces the Directly RTM widget's `status` into the Redux state tree. The purpose is to allow components to respond to changes in Directly's status. For example in this PR, on the `/help/contact` page we're now falling back to forum support if Directly fails to load.

### To test

Sign in as a user with no paid upgrades and head to http://calypso.localhost:3000/help/contact. You should see the form with a submit button saying "Ask an Expert" (this is the Directly integration form).

Open Redux Dev Tools. You should be able to see both a `DIRECTLY_INITIALIZATION_START` and a `DIRECTLY_INITIALIZATION_SUCCESS` action dispatched. You should also see the current status under `state.help.directly.status`.

**To test a failed lib load** change [the library URL in `lib/directly`](https://github.com/Automattic/wp-calypso/blob/master/client/lib/directly/index.js#L18) to a bogus address. Refresh the page and you should see the forums form appear instead. Redux Dev Tools should show a `DIRECTLY_INITIALIZATION_START` and a `DIRECTLY_INITIALIZATION_ERROR` action, and the state tree should reflect the error.

### CODE REVIEW

File: client/state/data-layer/third-party/directly/index.js [hunk 1]
```diff
@@ -3,21 +3,28 @@
...
 
 export function askQuestion( store, action, next ) {
 	directly.askQuestion( action.questionText, action.name, action.email );
 	next( action );
 }
 
-export function initialize( store, action, next ) {
-	directly.initialize();
+export function initialize( { dispatch }, action, next ) {
+	directly.initialize().then(
```
Comment (omarjackman) [inline client/state/data-layer/third-party/directly/index.js]: This is probably a matter of preference but I prefer the `.then( ... ).catch( ... )` structure.

File: client/me/help/help-contact/index.jsx [hunk 2]
```diff
@@ -581,7 +587,9 @@ const HelpContact = React.createClass( {
 		const olarkReadyOrTimedOut = olark.isOlarkReady || this.props.olarkTimedOut;
 		const ticketReadyOrError = ticketSupportConfigurationReady || null != ticketSupportRequestError;
 
-		return ! sitesInitialized || ! ticketReadyOrError || ! olarkReadyOrTimedOut;
+		const waitingOnDirectly = this.shouldUseDirectly() && this.props.directlyStatus !== DIRECTLY_STATUS_READY;
```
Comment (omarjackman) [inline client/me/help/help-contact/index.jsx]: Maybe add another selector for this called `isDirectlyReady` so we don't need to check for these statuses.

File: client/me/help/help-contact/index.jsx [hunk 3]
```diff
@@ -436,7 +438,11 @@ const HelpContact = React.createClass( {
 
 	shouldUseDirectly: function() {
 		const isEn = this.props.currentUserLocale === 'en';
-		return config.isEnabled( 'help/directly' ) && isEn;
+		return (
+			config.isEnabled( 'help/directly' ) &&
+			isEn &&
+			this.props.directlyStatus !== DIRECTLY_STATUS_ERROR
```
Comment (omarjackman) [inline client/me/help/help-contact/index.jsx]: Maybe we could create a selector for this called `isDirectlyFailed` that way we don't need to add case logic for the status in our component.

File: client/state/action-types.js [hunk 4]
```diff
@@ -67,7 +67,9 @@ export const CURRENT_USER_FLAGS_RECEIVE = 'CURRENT_USER_FLAGS_RECEIVE';
 export const CURRENT_USER_ID_SET = 'CURRENT_USER_ID_SET';
 export const DESERIALIZE = 'DESERIALIZE';
 export const DIRECTLY_ASK_QUESTION = 'DIRECTLY_ASK_QUESTION';
-export const DIRECTLY_INITIALIZE = 'DIRECTLY_INITIALIZE';
+export const DIRECTLY_INITIALIZATION_START = 'DIRECTLY_INITIALIZATION_START';
+export const DIRECTLY_INITIALIZATION_SUCCESS = 'DIRECTLY_INITIALIZATION_SUCCESS';
+export const DIRECTLY_INITIALIZATION_ERROR = 'DIRECTLY_INITIALIZATION_ERROR';
```
Comment (mattwondra) [inline client/state/action-types.js]: Ha — my brain is thinking "state machine" so naturally `START` should come first right??? :P

**Comment by mattwondra** [comment]
@omarjackman I tried implementing your Promise suggestions and my tests started failing unexpectedly. It seems like the order of `catch` cases was not matching what I expected. I did some research (since I haven't worked this heavily with Promises before) and it turns out there's some subtle but important gotchas to note when using `then` and `catch`.

When using `myPromise.then(...)` a new `Promise` is created and returned. So if you do `myPromise.then(...).catch(...)` you're actually catching errors in the `then` promise, not in `myPromise` itself. If `then` does not have an error handler, it will pass the error through to the `catch` but all of `myPromise`'s handlers will be called before the chained `catch` is called. This example illustrates how this can be deceiving:

```js
// Set up Promise and expose reject method
let rejectPromise;
let myPromise = new Promise( ( resolve, reject ) => rejectPromise = reject );

// Set up Promise handlers
myPromise.then(
  () => console.log( 'RESOLVED' )
).catch(
  () => console.log( 'REJECTED' )
);

myPromise.catch( () => console.log( 'ALL DONE' ) );

// Reject the promise to execute console logs
rejectPromise();
```

You'd expect the output to be:

```
REJECTED
ALL DONE
```

But it's not — it's:

```
ALL DONE
REJECTED
```

Because`myPromise` resolves all its direct handlers first, the `ALL DONE` handler is executed before the `then` falls through to the chained `REJECTED` handler.

So while these are equivalent:
```js
promise.then( null, errorHandler )
// same as
promise.catch( errorHandler )
```

And these are equivalent:
```js
promise.then( resolveHandler, errorHandler );
// same as
promise.then( resolveHandler );
promise.catch( errorHandler );
```

These are actually not equivalent and using one over the other is more than a matter of preference:
```js
promise.then( resolveHandler, errorHandler );
// not same as
promise.then( resolveHandler ).catch( errorHandler );
```

TIL! Thanks for opening this rabbit hole to me.

**Comment by mattwondra** [comment]
@omarjackman Ready for another peek and hopefully a :+1:

File: client/state/data-layer/third-party/directly/test/index.js [hunk 5]
```diff
@@ -60,5 +75,21 @@ describe( 'Directly data layer', () => {
...
+				.then( () => expect( store.dispatch ).to.have.been.calledWith( { type: DIRECTLY_INITIALIZATION_SUCCESS } ) )
+				.then( () => done() );
+
+			simulateInitializationSuccess();
+		} );
+
+		it( 'should dispatch an error action if initialization fails', ( done ) => {
+			initialize( store, action, next )
+				.then( () => expect( store.dispatch ).to.have.been.calledWith( { type: DIRECTLY_INITIALIZATION_ERROR } ) )
+				.then( done() );
```
Comment (omarjackman) [inline client/state/data-layer/third-party/directly/test/index.js]: Should be `() => done()`?

Comment (mattwondra) [inline client/state/data-layer/third-party/directly/test/index.js]: Good catch, fixing.

**Pull request merged**